jairus-m/dagster-sdlc
# Configure automated CI checks / pre-commit hooks + add unit tests

# Summary
Earlier this year, I set up pre-commits/slim CI for my work dbt project which focused on SQL code (`sqlfluff`), dbt compilation, other custom hooks, and `dbt build` validation. However, no additional custom CI past pre-commits were created.

I've always wanted to do this for a Python project that implemented both pre-commits and additional CI checks within github actions (that aren't just about the deployment validity/success) in order to follow best-practices with writing Python code and automating tests/checks in my development workflow. 

A few weeks ago, I came across this [repo](https://github.com/timvancann/yt-python-ci) that implements Astral's `uv` and `ruff` packages to setup a modern CI workflow to run checks on Python projects. The checks implemented here are based off that work but modified to fit the needs of my project. 

Thank you @timvancann for publishing that repo and that great YT video!
- [Setup a modern Python Code Quality CI Pipeline using UV, Ruff and more](https://www.youtube.com/watch?v=Y6D2XaFV3Cc)

# Details
Added two layers of automated checks:
- Pre-commit checks (on commit)
- CI checks (non-deployment, on pull request)

Added two unit test files in `dagster_proj_tests`:
- `tests_definitions.py`
-  `test_utils.py`

## Pre-Commit Hooks

```yml
default_install_hook_types: [pre-commit]
default_stages: [pre-commit]

repos:
  - repo: https://github.com/astral-sh/uv-pre-commit
    rev: 0.5.4
    hooks:
      - id: uv-lock
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.7.1
    hooks:
      - id: ruff
        args: [ --fix ]
      - id: ruff-format
  - repo: https://github.com/astral-sh/uv-pre-commit
    rev: 0.5.4
    hooks:
      - id: pip-compile
        args: [pyproject.toml, -o, requirements.txt]
        files: ^pyproject\.toml$
  - repo: local
    hooks:
      - id: pytest
        name: pytest
        entry: bash -c 'source .env && ./.venv/bin/pytest dagster_proj_tests/test_definitions.py'
        language: system
        pass_filenames: false
        always_run: true
```

This pre-commit config will:
- Create and lock the dependencies into `uv.lock`
- Use `ruff` to format any Python code
- Automatically update the `requirements.txt` if the `pyproject.toml` has been modified
- Run the `test_definitions.py` unit tests

## CI Workflow

```yml
name: uv ci
on: [pull_request]
jobs:
  lock_file:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/actions/setup
      - run: uv lock --locked 
  linting:
      runs-on: ubuntu-latest
      needs: [lock_file]
      steps:
        - uses: actions/checkout@v4
        - uses: ./.github/actions/setup
        - run: uvx ruff check .
  formatting:
    runs-on: ubuntu-latest
    needs: [lock_file]
    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/actions/setup
      - run: uvx ruff format --check .
  tests:
    runs-on: ubuntu-latest
    needs: [lock_file]
    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/actions/setup
      - run: uv add --dev pytest
      - run: uv run pytest --disable-warnings # run unit tests without warnings
  build:
    runs-on: ubuntu-latest
    needs: [linting, formatting, tests]
    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/actions/setup
      - run: uv build # build the package
```

This Github Action will:
- Check to make sure that the `uv.lock` file is in sync with the `proproject.yml file`
- Lint all python code in the directory 
- Output files that need formatting
- Run unit tests
- Build the Python package
